MAGIC WP4 and WP6 hyint

earnone opened the pull request:
Refined for v.2.0 and available for demo. Should work on 1975-2100 period and daily data so it takes too long for real time calculation.

TODO: It keeps a separate file with parameter settings (`hyint_parameters.R`) because of difficulties in introducing matrices into yml.

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -0,0 +1,85 @@
+
+preprocessors:
+  preproc:
+    extract_levels: false
```
There is no need to disable this preprocessor function, it is disabled by default

earnone replied:
updated

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -0,0 +1,85 @@
+        # a) false to keep original resolution
+        # b) set desired regridding resolution in cdo format e.g., "r320x160"
+        # c) "REF" to use resolution of reference model
+        rgrid: false   
```
This should preferably be done with the preprocessor, is there a good reason why cdo is used now?

earnone replied:
We first need to evaluate climate indices on the original grid, then regrid on a common grid if required (e.g. to produce difference of fields from two models). So the preproc regridding cannot be used.

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -0,0 +1,85 @@
+
+        ## ------- Advanced settings ----------
+        
+        topography_highres: "~/work/data/Elevation/GMTED2010_15n030_0125deg.nc"
```
Is this file needed? If so, how do I obtain it?

earnone replied:
This option was removed for standard users. We may later on include this option by passing a high resolution topography through the preproc

earnone commented:
Some final changes are ongoing to simplify the code. Documentation and provenance were added.

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -37,6 +38,12 @@ diagnostics:
       main:
         script: hyint/hyint.R
+        # General
+        run_diagnostic: true
```
Most of these options should be set in [config-user.yml](https://github.com/ESMValGroup/ESMValTool/blob/version2_development/esmvaltool/config-user.yml) instead of in the recipe.

earnone replied:
Do you mean run_diagnostic, write_netcdf, write_plots? OK, but something like write_plots is difficult to remove from here. How can a user call two diagnostics requiring only parts of their functionalities? E.g. when needing only the netcdf file from a first diagnostic (which would require write_plots: false) and using a second one to do a further processing with plots (which would require write_plots: true)? This would be the case when running the extremeEvents to produce netcdf with climdex indices (avoiding plots) followed hyint to produce joint analysis of hyint indices and climdex indices. 

Forcing these diagnostic-dependent parameters in a single configuration file is not user friendly. Why not allowing to reassign these parameters in recipes for special cases? The same applies to the path to data.

bouweandela replied:
> Do you mean run_diagnostic, write_netcdf, write_plots? 

Yes

> How can a user call two diagnostics requiring only parts of their functionalities? E.g. when needing only the netcdf file from a first diagnostic (which would require write_plots: false) and using a second one to do a further processing with plots (which would require write_plots: true)?

I agree that it could make sense in that case, though I do not think it is a problem if you get some extra plots out of a diagnostic run.

> Forcing these diagnostic-dependent parameters in a single configuration file is not user friendly. Why not allowing to reassign these parameters in recipes for special cases?

I think we need a careful look at these configuration options at some point, i.e. if they are necessary at all, but maybe not now. They add complication to diagnostic scripts and many diagnostic scripts do not even implement them.

The option `run_diagnostic` doesn't make much sense in a recipe, you can just comment out the diagnostic that you do not want to run or if you only want to run some of the diagnostics from a recipe we have a command line option for that.

> The same applies to the path to data.

Which path to the data?

earnone replied:
> Which path to the data?
OBS in the config-user.yml could be more handy for the local user if specified directly in the recipe as one may have data in different locations

mattiarighi replied:
The recipe should remain 100% portable, therefore all path must be defined in `config-user.yml`.

bouweandela replied:
> one may have data in different locations

Note that it is possible to specify multiple data locations in config-user.yml, e.g.

```yaml
rootpath:
  OBS: [~/obs_inputpath1, ~/obs_inputpath2]
```

The reason that this is done in a single config-user.yml file instead of in the recipe is that we want to be able to run all recipes with just one user configuration file. Otherwise you would need to adapt every recipe to your local paths before you can run it.

earnone replied:
OK, the multiple path option makes it much more handy, thanks.

bouweandela commented on esmvaltool/recipes/recipe_hyintex.yml:
```diff
@@ -37,6 +38,12 @@ diagnostics:
       main:
         script: hyint/hyint.R
+        # General
+        run_diagnostic: true
```
Most of these options should be set in [config-user.yml](https://github.com/ESMValGroup/ESMValTool/blob/version2_development/esmvaltool/config-user.yml) instead of in the recipe.

bouweandela commented on esmvaltool/recipes/recipe_hyintex.yml:
```diff
@@ -0,0 +1,130 @@
+        # Specify external path to repository of ETCCDI indices 
+        # (unless these are calculated in the same directory
+        # calling the extremeEvents diagnostic)
+        etccdi_dir <- ""
```
This is not valid yaml.

earnone replied:
Typo, ok

bouweandela commented on esmvaltool/recipes/recipe_hyintex.yml:
```diff
@@ -0,0 +1,130 @@
+# recipe_hyint.yml
```
This looks like a duplicate of recipe_hyint.yml. Is this file here intentionally?

earnone replied:
Intentionally. This is a special case of the main hyint recipe, in this case reading climate extreme indices produced by the climdex library with the extremeEvents recipe/namelist/diagnostics (which should then work one in line with the other one). We are awaiting further progress on the porting of the extremeEvents diagnostics into v2, meanwhile attempting to produce the indices ourself. The climdex library is not able to deal with netcdf files pre-processed by the v2 preprocessor. I will (re-?)open an issue on the climdex library.

bouweandela replied:
Ok, thank you for the explanation.

earnone replied:
I removed this recipe until it can be used to avoid confusion.

bouweandela commented:
The R dependencies have not been added to the file esmvaltool/install/R/r_requirements.txt, so when I try to run the diagnostic script it crashes with the following error message:
```
Error in library(ncdf4) : there is no package called ‘ncdf4’
Execution halted
```

earnone commented:
Please wait  before new tests for a new version to be uploaded shortly

bouweandela commented:
Tested again on Jasmin, mostly good results, but recipe option `selregions: [4]` and `selregions: [7]` crash with
```
in/hyint_ACCESS1-0_historical-rcp85_r1i1p1_2000_2010_ALL_tseries_.nc"
[1] "hyint: starting figures"
[1] "No value for range: assigning min and max"

 *** caught bus error ***
address 0x7f5e6b593090, cause 'non-existent physical address'

Traceback:
 1: axis(side = side, at = at, labels = labels, ...)
 2: Axis.default(...)
 3: Axis(...)
 4: localAxis(if (is.null(y)) xy$x else x, side = 1, ...)
 5: plot.default(time, type = "n", ylim = c(tmp.levels[1], tmp.levels[length(tmp.levels)]),     xlim = xlim, xlab = "Year", ylab = ylab, main = title_unit_m[ifield,         3])
 6: plot(time, type = "n", ylim = c(tmp.levels[1], tmp.levels[length(tmp.levels)]),     xlim = xlim, xlab = "Year", ylab = ylab, main = title_unit_m[ifield,         3])
 7: hyint_plot_trends(work_dir, plot_dir, ref_idx, seas)
An irrecoverable exception occurred. R is aborting now ...
```
and `selregions: [11]` crashes with
```
[1] "Working on  NA"
Error in regions[iselreg, ] : subscript out of bounds
Calls: hyint_trends -> calc_region_timeseries -> which
Execution halted
```

earnone commented:
Ok, looking into this, thanks.

Il Lun 25 Feb 2019, 21:13 Bouwe Andela <notifications@github.com> ha
scritto:

> Tested again on Jasmin, mostly good results, but recipe option selregions:
> [4] and selregions: [7] crash with
>
> in/hyint_ACCESS1-0_historical-rcp85_r1i1p1_2000_2010_ALL_tseries_.nc"
> [1] "hyint: starting figures"
> [1] "No value for range: assigning min and max"
>
>  *** caught bus error ***
> address 0x7f5e6b593090, cause 'non-existent physical address'
>
> Traceback:
>  1: axis(side = side, at = at, labels = labels, ...)
>  2: Axis.default(...)
>  3: Axis(...)
>  4: localAxis(if (is.null(y)) xy$x else x, side = 1, ...)
>  5: plot.default(time, type = "n", ylim = c(tmp.levels[1], tmp.levels[length(tmp.levels)]),     xlim = xlim, xlab = "Year", ylab = ylab, main = title_unit_m[ifield,         3])
>  6: plot(time, type = "n", ylim = c(tmp.levels[1], tmp.levels[length(tmp.levels)]),     xlim = xlim, xlab = "Year", ylab = ylab, main = title_unit_m[ifield,         3])
>  7: hyint_plot_trends(work_dir, plot_dir, ref_idx, seas)
> An irrecoverable exception occurred. R is aborting now ...
>
> and selregions: [11] crashes with
>
> [1] "Working on  NA"
> Error in regions[iselreg, ] : subscript out of bounds
> Calls: hyint_trends -> calc_region_timeseries -> which
> Execution halted
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/754#issuecomment-467165163>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AblrsW5EHfCk6E7CPTn8A-wCY3qO-iIVks5vREPWgaJpZM4ZMeu2>
> .
>

earnone commented:
I have added a check on user settings wrongly assigned user settings for selregions, selfields and plot_type. This resolves error for selregions: [11].

However, I have tried various settings with selregions: [4, 7] but it never crashed. Can you share the full recipe you used?

Thanks

> Tested again on Jasmin, mostly good results, but recipe option `selregions: [4]` and `selregions: [7]` crash with
> 
> ```
> in/hyint_ACCESS1-0_historical-rcp85_r1i1p1_2000_2010_ALL_tseries_.nc"
> [1] "hyint: starting figures"
> [1] "No value for range: assigning min and max"
> 
>  *** caught bus error ***
> address 0x7f5e6b593090, cause 'non-existent physical address'
> 
> Traceback:
>  1: axis(side = side, at = at, labels = labels, ...)
>  2: Axis.default(...)
>  3: Axis(...)
>  4: localAxis(if (is.null(y)) xy$x else x, side = 1, ...)
>  5: plot.default(time, type = "n", ylim = c(tmp.levels[1], tmp.levels[length(tmp.levels)]),     xlim = xlim, xlab = "Year", ylab = ylab, main = title_unit_m[ifield,         3])
>  6: plot(time, type = "n", ylim = c(tmp.levels[1], tmp.levels[length(tmp.levels)]),     xlim = xlim, xlab = "Year", ylab = ylab, main = title_unit_m[ifield,         3])
>  7: hyint_plot_trends(work_dir, plot_dir, ref_idx, seas)
> An irrecoverable exception occurred. R is aborting now ...
> ```
> and `selregions: [11]` crashes with
> 
> ```
> [1] "Working on  NA"
> Error in regions[iselreg, ] : subscript out of bounds
> Calls: hyint_trends -> calc_region_timeseries -> which
> Execution halted
> ```

bouweandela commented:
> I have added a check on wrongly assigned user settings for selregions, selfields and plot_type. This resolves error for selregions: [11].

Can you please push it to GitHub so I can test?

> However, I have tried various settings with selregions: [4, 7] but it never crashed. Can you share the full recipe you used?

I just ran the recipe again with `selregions: [4]` (all other settings the same as the default) and it completed successfully, so this was probably some issue with the Jasmin cluster.

earnone commented:
Version with checks on input parameters uploaded. This version includes also several minor refinements to the figures. I can also implement system2 as done for the quantilebias if we have time.

bouweandela commented:
> I can also implement system2 as done for the quantilebias if we have time.

It would be great if you could do it with the same function as implemented there

bouweandela commented on esmvaltool/diag_scripts/hyint/hyint_functions.R:
```diff
@@ -0,0 +1,1341 @@
+
+cdo <- function(command, args="", input="", options="", output="") {
+  if (args != "") args <- paste0(",", args)
+  argstr <- paste0(options, " ", command, args, " ", input, " ", output)
```
The input and output files need to be quoted, in case there are e.g. spaces or other special characters somewhere in the path.
```suggestion
  argstr <- paste0(options, " ", command, args, " '", input, "' '", output, "'")
```

earnone commented:
7b74721 is failing because input is at time a string containing further arguments in `ncdf_opener_universal`:

`   if (interp2grid) {
      print(paste("Remapping with CDO on", grid, "grid"))
      if (is.null(namevar)) {
        namefile <- cdo(remap_method, args = grid, input = namefile,
                        output = tempfile())
      } else {
        namefile <- cdo(remap_method, args = grid,
                        input = paste0("-selvar,", namevar, " ", namefile),
                        output = tempfile())
      }
    }
`
Looking at solution.

jhardenberg commented:
Yes, to be expected, see my comment https://github.com/ESMValGroup/ESMValTool/pull/750#issuecomment-469326419

earnone commented:
From our point of view this is ready to be merged. If needed, we will apply further refinements in a separate branch.

rvanharen commented:
The regridding for field_exp in hyint_plot_maps.R tries to regrid to a netCDF file and not to a file with grid information:
https://github.com/ESMValGroup/ESMValTool/blob/b9466958cc1c9be0ecd4d52bd03f6b32f547a9c7/esmvaltool/diag_scripts/hyint/hyint_plot_maps.R#L136-L138

I suggest to replace the regridding of field_exp with:
```
# regrid when comparing
ref_filename_grid <- paste0(tools::file_path_sans_ext(ref_filename), ".grid")
cdo("griddes", input = ref_filename, stdout = ref_filename_grid)
field_exp <- ncdf_opener(filename, field, "lon", "lat", rotate = "no",
                         interp2grid = T, grid = ref_filename_grid)
```

rvanharen commented:
After regridding the field_exp, I end up with a bunch of empty plots where field_exp is used for creating the map plot. I'm not sure yet where it is going wrong, but it appears that the field_exp variable ends up having no data in it.

jhardenberg commented:
> The regridding for field_exp in hyint_plot_maps.R tries to regrid to a netCDF file and not to a file with grid information:
Using a reference netcdf file to specify grid information is valid for cdo (it will regrid to match the grid of the reference file).
I do not quite understand, did you test the recipe and it did not work for you? With which settings ?

rvanharen commented:
Sorry didn't know that was valid cdo syntax. Our test script creates spaces in the path for some settings, which both the diagnostic script and cdo don't really like, so was working around that. 
Would there be any reason why the syntax I suggested would result in empty field_exp and the original not (I'll test it on jasmin by using the original syntax with relative paths of that produces valid field_exp plots)?

rvanharen commented:
I changed the ncdf_opener calls in hyint_plot_maps.R to use relative paths locally and used the netCDF file for the remapping, but I'm still running into the empty plot issue. For example:
```
[1] "/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004, 2006]_20190305_110034/plots/hyint/main/dsl_norm_CCSM4_historical-rcp85_r1i1p1_2000-2010_ALL_myearmean_EU_map.png"
[1] "region:  East-Asia"
[1] "hyint: plotting data for  East-Asia-2000-2010"
[1] "No value for range: assigning min and max"
```
and then finally it crashes in the code where the plot levels are calculated from tmp_field_max because tmp_field_max evaluates to Inf:
```
Error in seq.default(-tmp_field_max, tmp_field_max, len = nlev) : 
  'from' must be a finite number
Calls: hyint_plot_maps -> seq -> seq.default
```
If you have ideas that would help debugging that would be helpful

earnone commented:
OK, thanks, I now understand you have issues we cannot see.

I'll try to reproduce your situations and see what happens.

On Wed, 6 Mar 2019 at 11:58, Ronald van Haren <notifications@github.com>
wrote:

> I changed the ncdf_opener calls in hyint_plot_maps.R to use relative paths
> locally and used the netCDF file for the remapping, but I'm still running
> into the empty plot issue. For example:
>
> [1] "/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004, 2006]_20190305_110034/plots/hyint/main/dsl_norm_CCSM4_historical-rcp85_r1i1p1_2000-2010_ALL_myearmean_EU_map.png"
> [1] "region:  East-Asia"
> [1] "hyint: plotting data for  East-Asia-2000-2010"
> [1] "No value for range: assigning min and max"
>
> and then finally it crashes in the code where the plot levels are
> calculated from tmp_field_max because tmp_field_max evaluates to Inf:
>
> Error in seq.default(-tmp_field_max, tmp_field_max, len = nlev) :
>   'from' must be a finite number
> Calls: hyint_plot_maps -> seq -> seq.default
>
> If you have ideas that would help debugging that would be helpful
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/754#issuecomment-470064303>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AblrsUEzP0aSb9s7SjFp-W9LQ8ATQn6Eks5vT59pgaJpZM4ZMeu2>
> .
>

jhardenberg commented:
Hi @rvanharen , in order for us to reproduce the issue, could you provide us the recipe created by your test script ? (actually spaces in a unix path do sound a bit problematic indeed)

earnone commented:
> I changed the ncdf_opener calls in hyint_plot_maps.R to use relative paths locally and used the netCDF file for the remapping, but I'm still running into the empty plot issue. For example:
> 
> ```
> [1] "/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004, 2006]_20190305_110034/plots/hyint/main/dsl_norm_CCSM4_historical-rcp85_r1i1p1_2000-2010_ALL_myearmean_EU_map.png"
> [1] "region:  East-Asia"
> [1] "hyint: plotting data for  East-Asia-2000-2010"
> [1] "No value for range: assigning min and max"
> ```
> and then finally it crashes in the code where the plot levels are calculated from tmp_field_max because tmp_field_max evaluates to Inf:
> 
> ```
> Error in seq.default(-tmp_field_max, tmp_field_max, len = nlev) : 
>   'from' must be a finite number
> Calls: hyint_plot_maps -> seq -> seq.default
> ```
> If you have ideas that would help debugging that would be helpful

Since

 `tmp_field_max <- max(abs(tmp_field), na.rm = T)`
you must have Inf in the `tmp_field`, with`tmp_field <- field_exp - field_ref`. Can you check the field_exp and field_ref soon after reading them? You can use `is.infinite(x)`. I see no issue in my data.

earnone commented:
Running with default setting (as I can partially guess from the filename
you quote) works fine. You can see our output here:
http://wilma.to.isac.cnr.it/arnone/temp/esmvaltool_output/recipe_hyint_20190306_112916/plots/hyint/main/

On Wed, 6 Mar 2019 at 12:09, Jost von Hardenberg <notifications@github.com>
wrote:

> Hi @rvanharen <https://github.com/rvanharen> , in order for us to
> reproduce the issue, could you provide us the recipe created by your test
> script ? (actually spaces in a unix path do sound a bit problematic indeed)
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/754#issuecomment-470067729>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AblrsaKtZcGW1Vuj6NLaoQwWN4LCh_o1ks5vT6H3gaJpZM4ZMeu2>
> .
>

rvanharen commented:
> Hi @rvanharen , in order for us to reproduce the issue, could you provide us the recipe created by your test script ? (actually spaces in a unix path do sound a bit problematic indeed)

```
add_trend: true
add_trend_sd: true
add_trend_sd_shade: true
add_tseries_lines: true
add_zeroline: true
autolevels: true
autolevels_scale: 1
autolevels_scale_t: 1
boxregion: false
input_files: ['/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004,
    2006]_20190305_110034/preproc/hyint/pr/metadata.yml']
lm_trend: true
log_level: info
max_data_filesize: 100
norm_years: [2004, 2006]
npancol: 2
npanrow: 3
oplot_grid: false
output_file_type: png
plot_dir: /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004,
  2006]_20190305_110034/plots/hyint/main
plot_type: [1, 2, 3, 11, 12, 13, 14, 15]
profile_diagnostic: false
recipe: recipe_hyint_norm_years_[2004, 2006].yml
removedesert: false
rgrid: false
run_dir: /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004,
  2006]_20190305_110034/run/hyint/main
scale100years: true
scalepercent: false
script: main
selfields: [8, 4, 1, 9, 3, 2]
selregions: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
trend_years: false
trend_years_only: false
version: 2.0a1
weight_tseries: true
work_dir: /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/output_dir/recipe_hyint_norm_years_[2004,
  2006]_20190305_110034/work/hyint/main
write_netcdf: true
write_plots: true
```

bouweandela commented:
> I see no issue in my data.

Could you please check with the CP4CDS dataset on Jasmin? Because that is the data that the diagnostic should work with for the MAGIC project.

jhardenberg commented:
@rvanharen what you list is the settings.yml file, does your test script produce directly those or rather a recipe*yml file ? what is in your metadata.yml file ? I would like to be able to reproduce you error before giving an opinion.

rvanharen commented:
@jhardenberg I did update the paste to the recipy shortly after I pasted the settings.yml

rvanharen commented:
> you must have Inf in the `tmp_field`, with`tmp_field <- field_exp - field_ref`. Can you check the field_exp and field_ref soon after reading them? You can use `is.infinite(x)`. I see no issue in my data.

After reading there is quite a bit of data in both field_exp and field_ref (btw I have no issues with field_ref, that one seems to contain data). I seem to 'lose' the data when the multiyear average is extracted:
```
          # MAPS: select required year (if requested, multiyear average
          #       is stored at iyear=1)
          print(paste0('field_exp max before multiyear average:', max(is.finite(field_exp), na.rm=T)))
          print(paste0('field_ref max before multiyear average:', max(is.finite(field_ref), na.rm=T)))

          field_ref <- field_ref[, , iyear]
          field_exp <- field_exp[, , iyear_ref]
          tmp_field <- field_exp
          print(paste0('field_exp max after multiyear average:', max(is.finite(field_exp), na.rm=T)))
          print(paste0('field_ref max after multiyear average:', max(is.finite(field_ref), na.rm=T)))
```
produces the following output:
```
[1] "field_exp max before multiyear average:1"
[1] "field_ref max before multiyear average:1"
[1] "field_exp max after multiyear average:0"
[1] "field_ref max after multiyear average:1"
[1] "No value for range: assigning min and max"

```

jhardenberg commented:
Hi, I just gave a look. The problem, as you mentioned comes from the fact that your script generates directory names like "recipe_hyint_plot_type_[2, 3, 12, 13, 14, 15]_20190306_12075" 
Rather than changing the diagnostic would it not be easier to fix the test script so that it avoids spaces in the directory names (this is a terrible idea anyway in my opinion ...) ? You found this problem in this diagnostic but I cannot guarantee that also other may not have objections to such funky directory names. The problem never came out before, because by default esmvaltool does not generate directory names with spaces (unless you call your diagnostic "hyint_plot_type_[2, 3, 12, 13, 14, 15]" of course ....)

rvanharen commented:
Spaces in the fullpath is IMO a valid use case. I already changed the plot_maps script to use relative paths, but I can do the same with the rest of the scripts and see if that solves the problem. 
In addition I will rerun the test using a path without spaces and report back if that solves it, but I imagine that something would have complained already as I have it locally single quoted or as relative path.

rvanharen commented:
I changed the recipy name to localtest.yml (so now there are no spaces in the path) and ran it again, but the issue is still the same on jasmin. This is the command I used to run it:
`esmvaltool -c /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/config_user.yml "/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-ronald/REFACTORING_hyint_local/localtest.yml" --skip-nonexistent
`

earnone commented:
OK, I reproduced the error. It is related to the selection of normalization years in an interval falling within the range of available years and not at the beginning as it is usually adopted. Will fix asap.

jhardenberg commented:
Ok guys, I implemented the same changes as in #750 for the cdo function. It now puts quotes around all filenames and so it now accepts also dirnames and filenames with spaces. I also followed @bouweandela 's suggestion and split all chained cdo commands into separate commands. Temp files are now generated automatically if no "output" is specified. 
It turns out that while most commands work, a few cdo commands (e.g. delete and copy) do not like input files with space (with quotes or without). Since we had only one of those I used a clumsy but effective fix (adding 0 to the file and copying it to a temporary file first) in that only case ....

@rvanharen could you please try again your test script with a fresh pull of this branch (without any of the later modifications which you were testing). It should simply work now.

earnone commented:
Quick comment on this: as already noted, cdo with separate commands may
take quite some time more to perform the requested operations so we should
then find a better way to deal with this.

On Thu, 7 Mar 2019 at 00:42, Jost von Hardenberg <notifications@github.com>
wrote:

> Ok guys, I implemented the same changes as in #750
> <https://github.com/ESMValGroup/ESMValTool/pull/750> for the cdo
> function. It now puts quotes around all filenames and so it now accepts
> also dirnames and filenames with spaces. I also followed @bouweandela
> <https://github.com/bouweandela> 's suggestion and split all chained cdo
> commands into separate commands. Temp files are now generated automatically
> if no "output" is specified.
> It turns out that while most commands work, a few cdo commands (e.g.
> delete and copy) do not like input files with space (with quotes or
> without). Since we had only one of those I used a clumsy but effective fix
> (adding 0 to the file and copying it to a temporary file first) in that
> only case ....
>
> @rvanharen <https://github.com/rvanharen> could you please try again your
> test script with a fresh pull of this branch (without any of the later
> modifications which you were testing). It should simply work now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ESMValGroup/ESMValTool/pull/754#issuecomment-470322459>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AblrsbmFdQPSNBzU5LsEPzfb_Vv9fCPtks5vUFJ_gaJpZM4ZMeu2>
> .
>

jhardenberg commented:
Hi @rvanharen where/when did you find this error? I tried the latest version of the hyint recipe using a diagnostic name in the recipe with spaces (which creates directory names with spaces) and it seems to work nicely. 
Also I tried, and for me 
`cdo remapcon,'my grid file.nc' input.nc out.nc `
works... (I tried both with cdo 1.9.3 and 1.9.6)

bouweandela commented:
The problem was apparently a comma in the filename..

jhardenberg commented:
Ah, ok, thank you, I managed indeed to reproduce the error choosing a name with spaces and commas for the diagnostic in the recipe. 
@rvanharen thank you for the suggestion with relative paths. Since the issue actually only appears when remapcon2 is used, I have chosen to implement an even simpler fix by doing a previous "cdo griddes". This now seems to work for me.

bouweandela commented on doc/sphinx/source/recipes/recipe_hyint.rst:
```diff
@@ -0,0 +1,128 @@
+Recipes are stored in recipes/
+ 
+* recipe_hyint.yml (evaluating the 6 hydroclimatic indices) 
+* recipe_hyintex.yml (also including a selection of ETCCDI indices)
```
This recipe does not appear to be included in this pull request. Can you add it?

earnone replied:
The second recipe is on hold because of the delay in porting extremeEvents in v2. Hyint reads the results from extremeEvents. Removed "recipe_hyintex.yml" line from doc.

bouweandela commented on doc/sphinx/source/recipes/recipe_hyint.rst:
```diff
@@ -0,0 +1,128 @@
+
+* selfields: indices to be analysed and plotted. Select one or more fields from the following list (order-sensitive) as a numerical index: 1=SDII, 2=DSL, 3=WSL, 4=HY-INT, 5=ABS_INT, 6=ABS_DSL, 7=ABS_WSL, 8=PA, 9=R95
+
+* selregions: Select regions for timeseries and maps from the following list as a numerical index: 1=World, 2=World60 (60S/60N), 3=Tropics (30S/30N), 4=South 5=America, 6=Africa, 7=North America, 8=India, 9=Europe, 10=East-Asia, 11=Australia
```
There is a mismatch between the documentation here and in the recipe. When I try to run the recipe, region 11 does not appear to exist(?). I think it would be much more user friendly and reliable to use names (i.e. strings) for identifying regions instead of these numbers.

earnone replied:
Typo in the numbering written in documentation, changed.

earnone replied:
Done

bouweandela commented on doc/sphinx/source/recipes/recipe_hyint.rst:
```diff
@@ -0,0 +1,128 @@
+
+* norm_years: first and last year of reference normalization period to be used for normalized indices
+
+* selfields: indices to be analysed and plotted. Select one or more fields from the following list (order-sensitive) as a numerical index: 1=SDII, 2=DSL, 3=WSL, 4=HY-INT, 5=ABS_INT, 6=ABS_DSL, 7=ABS_WSL, 8=PA, 9=R95
```
Why not allow using the names directly in the recipe? That would be much more user friendly.

earnone replied:
Done

bouweandela commented on doc/sphinx/source/recipes/recipe_hyint.rst:
```diff
@@ -0,0 +1,128 @@
+
+#. Data
+
+   * rgrid (false): Define whether model data should be regridded. (a) false to keep original resolution; (b) set desired regridding resolution in cdo format e.g., "r320x160"; (c) "REF" to use resolution of reference model
```
We have preprocessor function for this. Why not use that instead of cdo?

earnone replied:
Regridding needs to be applied after the calculation of metrics is performed.

bouweandela replied:
Ok

bouweandela commented on doc/sphinx/source/recipes/recipe_hyint.rst:
```diff
@@ -0,0 +1,128 @@
+#. Timeseries and trends
+
+   * weight_tseries (true): adopt area weights in timeseries
+   * trend_years (false): (a) F=all; (b) c(year1,year2) to apply trend calculation and plotting only to a limited time interval (year1<=years<=year2); (c) c(year1,year2,year3,year4) to apply trend to two separate time intervals (year1<=y's<=year2) and (year3<=y's<=year4)
```
This is not clear. What are the allowed values in yaml syntax?

earnone replied:
Refined text and removed unfriendly option with two different time intervals.

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -0,0 +1,121 @@
+# recipe_hyint.yml
```
Please use `yamllint` to clean up this file

bouweandela commented on doc/sphinx/source/recipes/recipe_hyint.rst:
```diff
@@ -0,0 +1,127 @@
+
+   * npancol (2): number of columns in timeseries/trends multipanel figures
+   * npanrow (3): number of rows in timeseries/trends multipanel figures
+   * autolevels (true): select automated (T) or pre-set (F) range of values in plots
```
Use yaml style `true` and `false` instead of R code

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -0,0 +1,121 @@
+    description: HyInt hydroclimatic indices calculation and plotting
+    variables:
+      pr:
+#       preprocessor: preproc
```
Why is this commented out?

bouweandela commented:
I realized that we were not testing all options in the documentation, so I've run some additional tests. There were two cases in which chaning an option from the default resulted in a crash:
- setting `lm_trend` to `false`: 
```
[1] "******** PLOT TYPE: 14 *********"
[1] 960 800
[1] "HyInt_trends: reading file  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_lm_trend_False_20190323_055259/work/hyint/main/hyint_ACCESS1-0_historical-rcp85_r1i1p1_2000_2010_ALL_tseries_.nc"
[1] "hyint: starting figures"
[1] "No value for range: assigning min and max"
[1] "No value for range: assigning min and max"
Error in plot.window(...) : need finite 'ylim' values
Calls: hyint_plot_trends -> plot -> plot.default -> localWindow -> plot.window
In addition: There were 19 warnings (use warnings() to see them)
Execution halted
```

- setting `removedesert` to `true`
```
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_removedesert_True_20190323_054118/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_2000_2010_ALL.nc"
Error in field_exp * retdes3D : non-conformable arrays
Calls: hyint_plot_maps
In addition: Warning messages:
1: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_removedesert_True_20190323_054118/plots/hyint/main' already exists
2: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_removedesert_True_20190323_054118/plots/hyint/main' already exists
3: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_removedesert_True_20190323_054118/plots/hyint/main' already exists
Execution halted
```

earnone commented:
Restored `removedesert` code lost during cleanup. Removed option `lm_trend`. Fixed a bug for calendar 360_day models. Testing again on all cmip5 models.

bouweandela commented:
Thank you for making the recipe options more user friendly. I have tested again and noticed that the issue with `removedesert: true` is still present:
```
[1] "cdo  remapcon2,/tmp/RtmpTMxPUl/file2d32563515239 '/tmp/RtmpTMxPUl/file2d3252b502a8b' /tmp/RtmpTMxPUl/file2d32521fb80bc "
cdo remapcon2: SCRIP second order conservative weights from gaussian (192x96) to lonlat (192x145) grid, with source mask (3516)
cdo remapcon2: Processed 755712 values from 1 variable over 41 timesteps [1.63s 3224MB]
 num [1:192, 1:145, 1:41] NA NA NA NA NA NA NA NA NA NA ...
NULL
 num [1:192, 1:96, 1:41] NA NA NA NA NA NA NA NA NA NA ...
NULL
Error in field_exp * exp_retdes3D : non-conformable arrays
Calls: hyint_plot_maps
In addition: Warning messages:
1: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_removedesert_True_20190329_170837/plots/hyint/main' already exists
2: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_removedesert_True_20190329_170837/plots/hyint/main' already exists
Execution halted
```
I also noticed that the memory required for running the recipe has increased a lot since the last time I tested.

earnone commented:
Thanks for spotting this. Fixed removedesert for plot_type=2 and rgrid=F option. More memory I guess because I increased the pre-set time interval to 1980-2020 to have a more meaningful example. If too large we can reduce it again.

bouweandela commented:
Thanks I will test again. The larger memory is no problem if it is as expected.

bouweandela commented:
It is still not working. Most tests fail with a message like this:
```
[1] "******** PLOT TYPE: 2 *********"
[1] "Reading reference  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_ACCESS1-0_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "Reading experiment  /group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/work/hyint/main/hyint_MPI-ESM-MR_historical-rcp85_r1i1p1_1980_2020_ALL.nc"
[1] "hyint: starting figures"
[1]  672 1260
[1] "region:  World"
[1] "hyint: plotting data for  World-1980-2020"
[1] "No value for range: assigning min and max"
Error in .filled.contour(as.double(x), as.double(y), z, as.double(levels),  : 
  dimension mismatch
Calls: hyint_plot_maps -> filled_contour3 -> .filled.contour
In addition: Warning messages:
1: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/plots/hyint/main' already exists
2: In dir.create(plot_dir_exp, recursive = T) :
  '/group_workspaces/jasmin2/cp4cds1/magic-test/tmp-bandela/REFACTORING_hyint/output_dir/recipe_hyint_autolevels_scale_t_1.5_20190418_115310/plots/hyint/main' already exists
Execution halted
```

Could you please test it a bit more thorough yourself and let me know when it works? You can find a list of the options that I try here: https://github.com/ESMValGroup/ESMValTool/blob/version2_recipe_settings_test_utility/esmvaltool/utils/testing/recipe_settings/options.yml#L242

earnone commented:
Restored missing code lines for regridding, added further checks on user options. Tested on all options, several models and combinations. We are working on adding provenance to figures and inclusion of data from the extremeEvents recipe in a separate branch.

bouweandela commented on esmvaltool/recipes/recipe_hyint.yml:
```diff
@@ -0,0 +1,114 @@
+        end_year: 2020
+        reference_dataset: "ACCESS1-0"
+        mip: day
+        field: T2Ds
```
field can be removed

bouweandela commented:
I tested with the latest updates, things have improved, but the issue with `cdo remapcon2` is back, see https://github.com/ESMValGroup/ESMValTool/pull/754#pullrequestreview-212809820 and https://github.com/ESMValGroup/ESMValTool/pull/754#issuecomment-471757393

earnone commented:
@bouweandela sorry, will update soon for remapcon2. We also have the provenance added to figures, I would push that too.

bouweandela commented:
Sure, that's fine.

bouweandela commented:
With the latest changes, setting the recipe option `rgrid: r100x50` fails:
```
[1] "hyint: starting routine"
[1] "hyint: pre-processing file: /var/scratch2/bandela/test/version2_development/output_dir/recipe_hyint_rgrid_r100x50_20190508_094403/preproc/hyint/pr/CMIP5_ACCESS1-0_day_historical-rcp85_r1i1p1_pr_1980-2020.nc"
[1] "cdo  griddes 'r100x50'   > '/tmp/RtmpGiJoHz/file73a171d76bf1'"
cdo griddes: Open failed on >r100x50<
No such file or directory
Error in cdo("griddes", input = rgrid, stdout = gridf) : 
  Failed ( 1 ): cdo  griddes 'r100x50'   > '/tmp/RtmpGiJoHz/file73a171d76bf1'
Calls: hyint_preproc -> cdo
Execution halted
```

earnone commented:
We are discussing whether to drop this artificial split of the recipe of WP4 and WP6, Andreas is ok. @nielsdrost, shall we drop it and keep only one tool for the portal too? Then the second separate recipe can be dropped and the different parameters in the recipes can be put together again.

bouweandela commented:
Hi Enrico,

Niels is on holiday, but I think it would be fine to make a single recipe if that is scientifically most meaningful. However, I would strongly prefer to merge this pull request before adding any new features. I'm almost done testing for the tenth time or so and it looks like recipe_hyint.yml is finally working now, so please don't push additional changes.

bouweandela commented:
Or is there already a substantial amount of code in this related to `recipe_hyint_trends.yml`? Because then I will stop testing now and you can let me know when everything works and I will test again.

bouweandela commented:
Ok, I tested recipe_hyint.yml and everything seems to work now, but see my questions above.

earnone commented:
OK, great. Yes, we should go ahead with keeping only one recipe, the split was artificial only for the workpackages in the project. This is actually just exactly as it was before, the second recipe I added was a user-friendly version calling only the timeseries and trends plots (i.e. plot_type = 11:15) with all regions selected. So there is no change at all to the code, nor to your testing (you already use all possibilities). I would just push a `recipe_hyint.yml` with default settings that include all plot_types values and remove the recipe_hyint_trends.yml. I will have to push then at least 10 drinks for you.
